Add EIP-5334: EIP-721 User And Expires And Level Extension (#5334)
ethereum/EIPs: EIPS/eip-5334.md

---
eip: 5334
title: EIP-721 User And Expires And Level Extension

> description: Add a time-limited role with restricted permissions to [EIP-721] tokens.

Reviewer: Markdown doesn't work in the preamble:

```suggestion
description: Add a time-limited role with restricted permissions to EIP-721 tokens.
```

Some projects already use this design scheme under different names such as “operator” or “controller” but as it becomes more and more prevalent, we need a unified standard to facilitate collaboration amongst all applications.

Furthermore, applications of this model (such as renting) often demand that user addresses have only temporary access to using the NFT. Normally, this means the owner needs to submit two on-chain transactions, one to list a new address as the new user role at the start of the duration and one to reclaim the user role at the end. This is inefficient in both labor and gas and so an “expires” and “level” function is introduced that would facilitate the automatic end of a usage term without the need of a second transaction.

## Specification

The keywords "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY" and "OPTIONAL" in this document are to be interpreted as described in RFC 2119.

### Contract Interface
Solidity Interface with NatSpec & OpenZeppelin v4 Interfaces (also available at [`IERC5334.sol`](../assets/eip-5334/IERC5334.sol)):

```solidity
interface IERC5334 {

>     // Logged when the user of a NFT is changed or expires and level is changed

Reviewer: Is this more correct?

```suggestion
    // Logged when the user of a NFT, expires, or level is changed
```

    /// @return The user address for this NFT
    function userOf(uint256 tokenId) external view returns(address);

    /// @notice Get the user expires of an NFT
    /// @dev The zero value indicates that there is no user
    /// @param tokenId The NFT to get the user expires for
    /// @return The user expires for this NFT
    function userExpires(uint256 tokenId) external view returns(uint256);

    /// @notice Get the user level of an NFT
    /// @dev The zero value indicates that there is no user
    /// @param tokenId The NFT to get the user level for
    /// @return The user level for this NFT
    function userLevel(uint256 tokenId) external view returns(uint256);
}
```

The `userOf(uint256 tokenId)` function MAY be implemented as `pure` or `view`.

The `userExpires(uint256 tokenId)` function MAY be implemented as `pure` or `view`.

The `userLevel(uint256 tokenId)` function MAY be implemented as `pure` or `view`.

The `setUser(uint256 tokenId, address user, uint64 expires)` function MAY be implemented as `public` or `external`.

The `UpdateUser` event MUST be emitted when a user address is changed or the user expires is changed or the user level is changed.

> The `supportsInterface` method MUST return `true` when called with `0xad092b5c`.

Reviewer 2: `0xad092b5c` is the interfaceId of [ERC-4907](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-4907.md),  this value should be changed.

## Rationale

This model is intended to facilitate easy implementation. Here are some of the problems that are solved by this standard:

### Clear Rights Assignment

> With Dual “owner” and “user” roles, it becomes significantly easier to manage what lenders and borrowers can and cannot do with the NFT (in other words, their rights). Additionally, owners can control who the user is and it’s easy for other projects to assign their own rights to either the owners or the users.

Reviewer: This belongs in the motivation section. Rationale should describe technical choices made within the EIP itself, but this subsection seems to be justifying the existence of the `user` role (which is the whole point of the EIP.)

### Simple On-chain Time Management

> Once a rental period is over, the user role needs to be reset and the “user” has to lose access to the right to use the NFT. This is usually accomplished with a second on-chain transaction but that is gas inefficient and can lead to complications because it’s imprecise. With the `expires` function, there is no need for another transaction because the “user” is invalidated automatically after the duration is over.

Reviewer: Similarly this section about `expires` is also Motivation.

### Easy Third-Party Integration

> In the spirit of permission less interoperability, this standard makes it easier for third-party protocols to manage NFT usage rights without permission from the NFT issuer or the NFT application. Once a project has adopted the additional `user` role and `expires` and `level`, any other project can directly interact with these features and implement their own type of transaction. For example, a PFP NFT using this standard can be integrated into both a rental platform where users can rent the NFT for 30 days AND, at the same time, a mortgage platform where users can use the NFT while eventually buying ownership of the NFT with installment payments. This would all be done without needing the permission of the original PFP project.

Reviewer: Again, I think this is more Motivation.

Things you might want to put in the Rationale section could be:

 - Why `setUser` is part of the specification instead of leaving it undefined.
 - Why `level` is a `uint8`.
 - Why `level` is not an optional extension.
 - Why you only support a single `user` instead of multiple.

---
eip: 5334

> title: 721 User And Expires And Level Extension

Reviewer 3: ```suggestion
title: ERC-721 User And Expires And Level Extension
```
Although I think this title can be improved - it isn't quite clear to me what is being proposed.

Author: ![image](https://user-images.githubusercontent.com/37472467/181148112-d2041b1e-2ad3-4a65-95f5-62c007ee3824.png)

> description: Add a time-limited role with restricted permissions to [EIP-721] tokens.

Reviewer 3: ```suggestion
description: Add a time-limited role with restricted permissions to ERC-721 tokens.
```

Author: ![image](https://user-images.githubusercontent.com/37472467/181148174-bd456816-b41e-4c18-910c-9e99ef74e4db.png)

Reviewer 4: See @SamWillsn's comment:
> Markdown doesn't work in the preamble:

```suggestion 
description: Add a time-limited role with restricted permissions to EIP-721 tokens.
```

### Easy Third-Party Integration

In the spirit of permission less interoperability, this standard makes it easier for third-party protocols to manage NFT usage rights without permission from the NFT issuer or the NFT application. Once a project has adopted the additional `user` role and `expires` and `level`, any other project can directly interact with these features and implement their own type of transaction. For example, a PFP NFT using this standard can be integrated into both a rental platform where users can rent the NFT for 30 days AND, at the same time, a mortgage platform where users can use the NFT while eventually buying ownership of the NFT with installment payments. This would all be done without needing the permission of the original PFP project.

## Backwards Compatibility

As mentioned in the specifications section, this standard can be fully EIP-721 compatible by adding an extension function set.

In addition, new functions introduced in this standard have many similarities with the existing functions in EIP-721. This allows developers to easily adopt the standard quickly.

## Reference Implementation

> [`ERC5334`] Implementation: [`ERC5334.sol`](../assets/eip-5334/ERC5334.sol)

Reviewer 3: This should either be whole implemented here in this section or in an asset file - I don't think it should be duplicated across both places.

Author: ![image](https://user-images.githubusercontent.com/37472467/181150723-5398dcb8-5d74-47f2-a06e-ccea0ac5fed4.png)

---
eip: 5334
title: 721 User And Expires And Level Extension
description: Add a time-limited role with restricted permissions to ERC-721 tokens.
author: Yan (@yan253319066)
discussions-to: https://ethereum-magicians.org/t/erc-721-user-and-expires-and-level-extension/10097
status: Draft
type: Standards Track
category: ERC
created: 2022-07-25
requires: 165, 721
---

## Abstract

> This standard is an extension of [EIP-721](./eip-721.md). It proposes an additional role (`user`) which can be granted to addresses, and a time where the role is automatically revoked (`expires`) and (`level`) . The `user` role represents permission to "use" the NFT, but not the ability to transfer it or set users.

Reviewer 5: EIPs don't propose, they assert.  Also, saying "this standard" is a bit redundant.
```suggestion
An [EIP-721](./eip-721.md) extension that adds an additional role (`user`) which can be granted to addresses, and a time where the role is automatically revoked (`expires`) and (`level`) . The `user` role represents permission to "use" the NFT, but not the ability to transfer it or set users.
```

Some projects already use this design scheme under different names such as “operator” or “controller” but as it becomes more and more prevalent, we need a unified standard to facilitate collaboration amongst all applications.

Furthermore, applications of this model (such as renting) often demand that user addresses have only temporary access to using the NFT. Normally, this means the owner needs to submit two on-chain transactions, one to list a new address as the new user role at the start of the duration and one to reclaim the user role at the end. This is inefficient in both labor and gas and so an “expires” and “level” function is introduced that would facilitate the automatic end of a usage term without the need of a second transaction.

## Specification

The keywords "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY" and "OPTIONAL" in this document are to be interpreted as described in RFC 2119.

### Contract Interface

> Solidity Interface with NatSpec & OpenZeppelin v4 Interfaces (also available at [IERC5334.sol](../assets/eip-5334/contracts/IERC5334.sol)):

Reviewer 5: @SamWilsn It would be nice if the bot didn't complain about this here:
> proposals must be referenced with the form `EIP-N` (not `ERC-N`)

Reviewer 4: Wrap it in code blocks
```suggestion
Solidity Interface with NatSpec & OpenZeppelin v4 Interfaces (also available at [`IERC5334.sol`](../assets/eip-5334/contracts/IERC5334.sol)):
```
